### ISSUE TITLE
consul: infinite retry on unknown DC

### ISSUE THREAD

**Issue by olix0r** [BuoyantIO/linkerd#647]
when an invalid/unknown DC is resolved through the consul namer, it keeps retrying this lookup indefinitely. I think this is because consul returns a 500 in this case

1. Is it appropriate to retry on all 5XX responses? What are the class of 5XX response that are legitimately retryable?
2. Can we use a less permissive retry policy on this call?
3. Does the recent rewrite help this (does the Activity release cause cancellation?) Should namerd eagerly release observations with no clients?  This might Just Fix this?  

reported by @leozc

**Comment by Ashald** [comment]
This is another tricky problem as there are now way to distinguish between temporarily disconnected datacenter and non existing one. So re-trying sounds legit.

OTOH, if we want to have a bulletproof solution we can think about an optional config field where we can list DCs that are known to exist for sure and enable re-try for them but fail for any other DCs.

**Comment by olix0r** [comment]
It's unfortunate that consul doesn't provide error codes that distinguish failure types.

Since this is more of a nuisance than an active problem, I don't think there's any urgency of having a bulletproof solution.

And, I think approach (3) is a more robust solution that will apply beyond the consul namer: I think it's fine for retries to occur as long as the name is being observed.  If we can solve observation lifecycle -- where we stop observing irrelevant names, the observation teardown should stop the retry behavior.

**Comment by adleong** [comment]
This may be getting off topic for this ticket, but there's some nuance to (3) because of the long polling interface that namerd exposes.  I think this could be accomplished by adding time-based eviction to the inactive observer cache.

**Comment by Ashald** [comment]
Oh yeah, being able to terminate observations based on some conditions would be great. :) And I agree, it would be better solution.

**Comment by olix0r** [comment]
Yes, that's absolutely right. I'd like to introduce a streaming (gRPC) api once our gRPC support lands, which may make this easier to reason about.

**Comment by Ashald** [comment]
Once this is done https://github.com/hashicorp/consul/issues/2736 I'd like to submit a PR to improve Consul namer in a way so that it'll check datacenter availability before fetching DC services - this would eliminate meaningless retries may DC outage occur. Does it sound good to you?
